File: js/ba64.test.ts
/** Property + chaos tests for ba64.ts (node:test, no external deps). */
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";
import zlib from "node:zlib";
import { Buffer } from "node:buffer";
import { encode, decode, Ba64Error } from "./ba64.ts";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const VECTORS = path.join(HERE, "..", "vectors");
const load = (n: string) => JSON.parse(readFileSync(path.join(VECTORS, n), "utf8")).cases;
const b64 = (b: Uint8Array) => Buffer.from(b).toString("base64");

// deterministic PRNG (mulberry32) so failures reproduce
function rng(seed: number) {
  return () => {
    seed |= 0; seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
function randBytes(r: () => number, n: number): Uint8Array {
  const a = new Uint8Array(n);
  for (let i = 0; i < n; i++) a[i] = Math.floor(r() * 256);
  return a;
}

test("conformance: all committed vectors, exact codes", () => {
  const hex = (b: Uint8Array) => Buffer.from(b).toString("hex");
  const unhex = (s: string) => new Uint8Array(Buffer.from(s, "hex"));
  const expectErr = (input: string, code: string, max?: number, alt: string[] = []) =>
    assert.throws(() => decode(input, max !== undefined ? { maxDecodedLen: max } : {}),
      (e: any) => e instanceof Ba64Error && [code, ...alt].includes(e.code), `${input} -> ${code}`);

  for (const c of load("decode_plain.json")) assert.equal(hex(decode(c.input)), c.output_hex);
  for (const c of load("decode_frames.json")) assert.equal(hex(decode(c.input)), c.output_hex);
  for (const c of load("decode_errors.json")) expectErr(c.input, c.error, undefined, c.error_alt ?? []);
  for (const c of load("bombs.json")) expectErr(c.input, c.error, c.max_decoded_len);
  for (const c of load("encode_props.json")) {
    const data = unhex(c.input_hex), enc = encode(data), plain = b64(data);
    assert.equal(hex(decode(enc)), c.input_hex);
    assert.ok(enc.length <= plain.length);
    if (c.props.includes("compressed")) assert.ok(enc.startsWith("="));
    if (c.props.includes("plain")) assert.equal(enc, plain);
  }
  for (const c of load("differential.json")) {
    if (c.output_hex !== undefined) assert.equal(hex(decode(c.input)), c.output_hex);
    else expectErr(c.input, c.error);
  }
  // level knob must reach zlib (not be dropped)
  const d = Buffer.from("The quick brown fox. ".repeat(60));
  assert.notEqual(encode(d, { level: 1 }), encode(d, { level: 9 }));
});

test("property: roundtrip", () => {
  const r = rng(0xba64);
  for (let i = 0; i < 3000; i++) {
    const kind = r();
    const data = kind < 0.5 ? randBytes(r, Math.floor(r() * 64))
      : new Uint8Array(Math.floor(r() * 300)).fill(Math.floor(r() * 3));
    assert.deepEqual(decode(encode(data)), data);
  }
});

test("property: floor + marker-iff-strictly-smaller", () => {
  const r = rng(1);
  for (let i = 0; i < 3000; i++) {
    const data = r() < 0.5 ? randBytes(r, Math.floor(r() * 80))
      : Buffer.from("ab ".repeat(Math.floor(r() * 60)));
    const enc = encode(data);
    const plain = b64(data);
    assert.ok(enc.length <= plain.length, "floor");
    assert.equal(enc.startsWith("="), enc.length < plain.length);
    if (!enc.startsWith("=")) assert.equal(enc, plain);
  }
});

test("property: decode never throws non-Ba64Error", () => {
  const r = rng(2);
  for (let i = 0; i < 5000; i++) {
    const n = Math.floor(r() * 40);
    let s = "";
    for (let j = 0; j < n; j++) s += String.fromCharCode(Math.floor(r() * 128));
    try { const out = decode(s); assert.ok(out instanceof Uint8Array); }
    catch (e) { assert.ok(e instanceof Ba64Error, `non-taxonomy throw: ${e}`); }
  }
});

function validStrings(): string[] {
  return [...load("decode_plain.json"), ...load("decode_frames.json")].map((c) => c.input);
}
const safe = (s: string) => {
  try { assert.ok(decode(s) instanceof Uint8Array); }
  catch (e) { assert.ok(e instanceof Ba64Error, `non-taxonomy throw: ${e}`); }
};

test("chaos: bit-flip sweep", () => {
  let total = 0;
  for (const s of validStrings()) {
    const bytes = Buffer.from(s, "latin1");
    for (let i = 0; i < bytes.length; i++)
      for (let bit = 0; bit < 8; bit++) {
        const m = Buffer.from(bytes); m[i] ^= 1 << bit;
        safe(m.toString("latin1")); total++;
      }
  }
  assert.ok(total > 2000);
});

test("chaos: truncation sweep", () => {
  for (const s of validStrings())
    for (let n = 0; n < s.length; n++) safe(s.slice(0, n));
});

test("chaos: frame corruption never silently wrong", () => {
  for (const c of load("decode_frames.json")) {
    const frame = Buffer.from(c.input.slice(1), "base64");
    const original = c.output_hex;
    for (let i = 0; i < frame.length; i++)
      for (let bit = 0; bit < 8; bit++) {
        const m = Buffer.from(frame); m[i] ^= 1 << bit;
        const text = "=" + m.toString("base64");
        try {
          const out = Buffer.from(decode(text)).toString("hex");
          assert.equal(out, original, "frame silently returned wrong bytes");
        } catch (e) { assert.ok(e instanceof Ba64Error); }
      }
  }
});

test("chaos: inflation cap bounds memory", () => {
  const inflated = 128 * 2 ** 20;
  const payload = zlib.deflateRawSync(Buffer.alloc(inflated), { level: 9 });
  const crc = Buffer.alloc(4); // wrong crc is fine; length check fires first
  const frame = Buffer.concat([Buffer.from([1, 1, 100 & 0x7f]), crc, payload]);
  // decoded_len = 100 (single LEB128 byte)
  const text = "=" + frame.toString("base64");
  const before = process.memoryUsage().rss;
  assert.throws(() => decode(text), (e: any) => e instanceof Ba64Error && e.code === "E_LENGTH_MISMATCH");
  const delta = process.memoryUsage().rss - before;
  assert.ok(delta < 32 * 2 ** 20, `cap not enforced: rss grew ${delta}`);
});

File: go/ba64_test.go
package ba64

import (
	"bytes"
	"compress/flate"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"hash/crc32"
	"math/rand"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

type vcase struct {
	Name          string   `json:"name"`
	Input         string   `json:"input"`
	OutputHex     *string  `json:"output_hex"`
	Error         string   `json:"error"`
	ErrorAlt      []string `json:"error_alt"`
	InputHex      string   `json:"input_hex"`
	Props         []string `json:"props"`
	MaxDecodedLen *int     `json:"max_decoded_len"`
}

func load(t *testing.T, name string) []vcase {
	t.Helper()
	b, err := os.ReadFile(filepath.Join("..", "vectors", name))
	if err != nil {
		t.Fatal(err)
	}
	var doc struct {
		Cases []vcase `json:"cases"`
	}
	if err := json.Unmarshal(b, &doc); err != nil {
		t.Fatal(err)
	}
	return doc.Cases
}

func codeOf(err error) string {
	if e, ok := err.(*Error); ok {
		return e.Code()
	}
	return "non-taxonomy: " + err.Error()
}

func expectBytes(t *testing.T, input, outHex, name string) {
	got, err := Decode(input)
	if err != nil {
		t.Errorf("%s: unexpected error %v", name, err)
		return
	}
	if hex.EncodeToString(got) != outHex {
		t.Errorf("%s: decode -> %s, want %s", name, hex.EncodeToString(got), outHex)
	}
}

func expectError(t *testing.T, input, code, name string, max int, alt ...string) {
	_, err := DecodeMax(input, max)
	if err == nil {
		t.Errorf("%s: expected error %s, got nil", name, code)
		return
	}
	c := codeOf(err)
	accepted := append([]string{code}, alt...)
	for _, a := range accepted {
		if c == a {
			return
		}
	}
	t.Errorf("%s: got %s, want one of %v", name, c, accepted)
}

func TestErrorShape(t *testing.T) {
	_, err := Decode("SGVsbG8") // bad padding
	e, ok := err.(*Error)
	if !ok || e.Code() != "E_BASE64" || e.Error() != "E_BASE64" {
		t.Fatalf("bad error shape: %v", err)
	}
}

func TestConformance(t *testing.T) {
	for _, c := range load(t, "decode_plain.json") {
		expectBytes(t, c.Input, *c.OutputHex, "plain/"+c.Name)
	}
	for _, c := range load(t, "decode_frames.json") {
		expectBytes(t, c.Input, *c.OutputHex, "frame/"+c.Name)
	}
	for _, c := range load(t, "decode_errors.json") {
		expectError(t, c.Input, c.Error, "err/"+c.Name, DefaultMaxDecodedLen, c.ErrorAlt...)
	}
	for _, c := range load(t, "bombs.json") {
		max := DefaultMaxDecodedLen
		if c.MaxDecodedLen != nil {
			max = *c.MaxDecodedLen
		}
		expectError(t, c.Input, c.Error, "bomb/"+c.Name, max)
	}
}

func TestEncodeProps(t *testing.T) {
	for _, c := range load(t, "encode_props.json") {
		data, _ := hex.DecodeString(c.InputHex)
		enc := Encode(data)
		plain := base64.StdEncoding.EncodeToString(data)
		got, err := Decode(enc)
		if err != nil || hex.EncodeToString(got) != c.InputHex {
			t.Errorf("%s: roundtrip failed", c.Name)
		}
		if len(enc) > len(plain) {
			t.Errorf("%s: floor violated", c.Name)
		}
		for _, p := range c.Props {
			if p == "compressed" && !strings.HasPrefix(enc, "=") {
				t.Errorf("%s: expected compressed", c.Name)
			}
			if p == "plain" && enc != plain {
				t.Errorf("%s: expected plain", c.Name)
			}
		}
	}
	// level knob must reach flate (not be dropped)
	d := bytes.Repeat([]byte("The quick brown fox. "), 60)
	if EncodeLevel(d, 1) == EncodeLevel(d, 9) {
		t.Error("level argument has no effect")
	}
}

func TestDifferential(t *testing.T) {
	for _, c := range load(t, "differential.json") {
		if c.OutputHex != nil {
			expectBytes(t, c.Input, *c.OutputHex, "diff/"+c.Name)
		} else {
			expectError(t, c.Input, c.Error, "diff/"+c.Name, DefaultMaxDecodedLen)
		}
	}
}

// --- property ---------------------------------------------------------------

func TestPropertyRoundtrip(t *testing.T) {
	r := rand.New(rand.NewSource(0xBA64))
	for i := 0; i < 5000; i++ {
		var data []byte
		if r.Intn(2) == 0 {
			data = make([]byte, r.Intn(64))
			r.Read(data)
		} else {
			data = bytes.Repeat([]byte{byte(r.Intn(3))}, r.Intn(300))
		}
		got, err := Decode(Encode(data))
		if err != nil || !bytes.Equal(got, data) {
			t.Fatalf("roundtrip failed for %x", data)
		}
	}
}

func TestPropertyFloorAndMarker(t *testing.T) {
	r := rand.New(rand.NewSource(1))
	for i := 0; i < 5000; i++ {
		var data []byte
		if r.Intn(2) == 0 {
			data = make([]byte, r.Intn(80))
			r.Read(data)
		} else {
			data = bytes.Repeat([]byte("ab "), r.Intn(60))
		}
		enc := Encode(data)
		plain := base64.StdEncoding.EncodeToString(data)
		if len(enc) > len(plain) {
			t.Fatalf("floor violated for %x", data)
		}
		if strings.HasPrefix(enc, "=") != (len(enc) < len(plain)) {
			t.Fatalf("marker-iff-smaller violated for %x", data)
		}
	}
}

func TestPropertyNoCrash(t *testing.T) {
	r := rand.New(rand.NewSource(2))
	for i := 0; i < 20000; i++ {
		b := make([]byte, r.Intn(40))
		for j := range b {
			b[j] = byte(r.Intn(128))
		}
		_, err := Decode(string(b))
		if err != nil {
			if _, ok := err.(*Error); !ok {
				t.Fatalf("non-taxonomy error: %v", err)
			}
		}
	}
}

// --- chaos ------------------------------------------------------------------

func validStrings(t *testing.T) []struct{ in, out string } {
	var v []struct{ in, out string }
	for _, c := range load(t, "decode_plain.json") {
		v = append(v, struct{ in, out string }{c.Input, *c.OutputHex})
	}
	for _, c := range load(t, "decode_frames.json") {
		v = append(v, struct{ in, out string }{c.Input, *c.OutputHex})
	}
	return v
}

func assertSafe(t *testing.T, s string) {
	_, err := Decode(s)
	if err != nil {
		if _, ok := err.(*Error); !ok {
			t.Fatalf("non-taxonomy error on %q: %v", s, err)
		}
	}
}

func TestChaosBitFlipSweep(t *testing.T) {
	total := 0
	for _, v := range validStrings(t) {
		b := []byte(v.in)
		for i := range b {
			for bit := 0; bit < 8; bit++ {
				m := append([]byte(nil), b...)
				m[i] ^= 1 << bit
				assertSafe(t, string(m))
				total++
			}
		}
	}
	if total < 2000 {
		t.Fatalf("only %d bit-flips", total)
	}
}

func TestChaosTruncationSweep(t *testing.T) {
	for _, v := range validStrings(t) {
		for n := 0; n < len(v.in); n++ {
			assertSafe(t, v.in[:n])
		}
	}
}

func TestChaosFrameCorruptionNeverSilent(t *testing.T) {
	for _, c := range load(t, "decode_frames.json") {
		frame, _ := base64.StdEncoding.DecodeString(c.Input[1:])
		for i := range frame {
			for bit := 0; bit < 8; bit++ {
				m := append([]byte(nil), frame...)
				m[i] ^= 1 << bit
				text := "=" + base64.StdEncoding.EncodeToString(m)
				got, err := Decode(text)
				if err != nil {
					if _, ok := err.(*Error); !ok {
						t.Fatalf("non-taxonomy error: %v", err)
					}
					continue
				}
				if hex.EncodeToString(got) != *c.OutputHex {
					t.Fatalf("frame silently returned wrong bytes")
				}
			}
		}
	}
}

func TestChaosInflationCapBoundsMemory(t *testing.T) {
	const inflated = 128 << 20
	var b bytes.Buffer
	w, _ := flate.NewWriter(&b, 9)
	_, _ = w.Write(make([]byte, inflated))
	_ = w.Close()
	var crc [4]byte
	frame := append([]byte{version, methodDeflateRaw, 100}, crc[:]...) // decoded_len=100
	frame = append(frame, b.Bytes()...)
	text := "=" + base64.StdEncoding.EncodeToString(frame)

	var m0, m1 runtime.MemStats
	runtime.GC()
	runtime.ReadMemStats(&m0)
	_, err := Decode(text)
	runtime.ReadMemStats(&m1)
	if codeOf(err) != "E_LENGTH_MISMATCH" {
		t.Fatalf("want E_LENGTH_MISMATCH, got %v", err)
	}
	if grew := int64(m1.TotalAlloc - m0.TotalAlloc); grew > 32<<20 {
		t.Fatalf("cap not enforced: allocated %d bytes", grew)
	}
}

// unused import guard
var _ = binary.LittleEndian
var _ = crc32.IEEE

// --- native fuzz harnesses (go test -fuzz) ----------------------------------

func FuzzDecode(f *testing.F) {
	for _, c := range load(&testing.T{}, "decode_errors.json") {
		f.Add(c.Input)
	}
	f.Fuzz(func(t *testing.T, s string) {
		_, err := Decode(s) // must return bytes or a taxonomy error, never panic
		if err != nil {
			if _, ok := err.(*Error); !ok {
				t.Fatalf("non-taxonomy error: %v", err)
			}
		}
	})
}

func FuzzRoundtrip(f *testing.F) {
	f.Add([]byte("hello"))
	f.Fuzz(func(t *testing.T, data []byte) {
		got, err := Decode(Encode(data))
		if err != nil || !bytes.Equal(got, data) {
			t.Fatalf("roundtrip broken for %x", data)
		}
		if len(Encode(data)) > len(base64.StdEncoding.EncodeToString(data)) {
			t.Fatalf("floor broken for %x", data)
		}
	})
}
